features/jest/client/src/main/java/org/opennms/features/jest/client/JestClientWithCircuitBreaker.java

```diff
@@ -0,0 +1,140 @@
+
+    private EventForwarder eventForwarder = null;
+    private ApplicationContext context;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        try {
+            setEventForwarder(context.getBean(EventForwarder.class));
+        } catch (BeansException e) {
+            LOG.error("Cannot get an EventForwarder in JestClientWithCircuitBreaker", e);
+        }
+    }
+
+    @Override
+    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+        this.context = applicationContext;
+    }
+
+    public JestClientWithCircuitBreaker(JestClient client, CircuitBreaker circuitBreaker) {
+        this.client = Objects.requireNonNull(client);
+        this.circuitBreaker = Objects.requireNonNull(circuitBreaker);
+    }
+
+    //there is a cyclic dependency JestClientWithCircuitBreaker -> EventForwarder -> EventToIndex -> JestClientWithCircuitBreaker
+    //to brake it EventForwarder will be initialized later
+    public void setEventForwarder(EventForwarder eventForwarder) {
+        this.eventForwarder = Objects.requireNonNull(eventForwarder);
+        circuitBreaker.getEventPublisher()
+                .onStateTransition(e -> {
+                    final String from = e.getStateTransition().getFromState().toString();
+                    final String to = e.getStateTransition().getToState().toString();
+                    LOG.info("CircuitBreaker with name {} changed state from {} to {}", circuitBreaker.getName(), from, to);
+                    // Send an event when the circuit breaker's state changes (only when eventForwarder has been initialized)
+                    if (this.eventForwarder != null) {
+                        final Event event = new EventBuilder(CIRCUIT_BREAKER_STATE_CHANGE_EVENT_UEI, JestClientWithCircuitBreaker.class.getCanonicalName())
+                                .addParam("name", circuitBreaker.getName())
+                                .addParam("fromState", from)
+                                .addParam("toState", to)
+                                .getEvent();
+                        this.eventForwarder.sendNow(event);
+                    }
+                });
+    }
+
+    @Override
+    public <T extends JestResult> T execute(Action<T> clientRequest) throws IOException {
+        try {
+            return circuitBreaker.decorateCheckedSupplier(() -> client.execute(clientRequest)).apply();
+        } catch (IOException e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    public <T extends JestResult> void executeAsync(Action<T> clientRequest, JestResultHandler<? super T> jestResultHandler) {
+        CompletableFuture.runAsync(() -> {
+            try {
+                jestResultHandler.completed(execute(clientRequest));
+            } catch (IOException e) {
+                jestResultHandler.failed(e);
+            }
```

Comment:
This seems bogus to me. We do not call `client.execAsync` here but build our own async wrapper. Can we just delegate to `execAsync` instead, but call it with a wrapper around `jestResultHandler` which drives the circuit breaker?

Reply:
The wrapper around `jestResultHandler` works when when the HTTP request already has been made. But we need to prevent unnecessary HTTP requests.

Reply:
I see the only improvement in specifying an executor in `runAsync` (if there is a need to use specific one).